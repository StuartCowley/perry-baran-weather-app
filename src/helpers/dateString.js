const dateString = (dateTime) => {
  return new Date(dateTime * 1000).toDateString();
};

export default dateString;
