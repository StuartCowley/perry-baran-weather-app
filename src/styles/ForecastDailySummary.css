.forecast-daily-summary {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 10px;
  width: 16%;
  padding-top: 5px;
  margin: 0 0.5%;
  border-radius: 4px;
  background-image: linear-gradient(-8deg, var(--dark-blue), var(--medium-blue));  
  text-align: center;
}

.forecast-daily-summary > p {
  font-size: min(1.2rem, 3cqw);
  color: var(--almost-white);
  text-shadow: 1px 1px 2px var(--almost-black);
}

.forecast-daily-summary__icon {
  font-size: min(4rem, 5cqw);
}

.forecast-daily-summary__more-details {
  all: unset;
  width: 80%;
  max-width: 120px;
  padding: 5px 0;
  border-radius: 4px;
  margin-bottom: 10px;
  border: 2px solid var(--almost-white);
  color: white;
  font-size: min(1rem, 1.7cqw);
  font-weight: 600;
  transition-duration: 100ms;
  cursor: pointer;
}

.forecast-daily-summary__more-details:disabled {
  opacity: 0.4;
}

.forecast-daily-summary__more-details:hover {
  background-color: var(--almost-white);
  color: var(--dark-blue);
}

@media (min-width: 900px) {
  .forecast-daily-summary br {
    display: none;
  }
}