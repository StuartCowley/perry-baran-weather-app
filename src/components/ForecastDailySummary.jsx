import React from "react";
import PropTypes from "prop-types";
import WeatherIcon from "react-icons-weather";
import "../styles/ForecastDailySummary.css";
import getUnits from "../helpers/getUnits";
import useUnitContext from "../hooks/useUnitContext";
import dateString from "../helpers/dateString";

function ForecastDailySummary({
  dateTime,
  weather,
  icon,
  temp,
  handleForecastSelect,
  selectedDate,
}) {
  const { units } = useUnitContext();
  const { tempUnits } = getUnits(units);
  const weatherDescription = weather ? weather.split(" ") : [weather];
  const date = dateString(dateTime);
  const dayOfTheWeek = date.slice(0, 3);
  const monthAndDay = date.slice(3, 10);

  return (
    <div
      className="forecast-daily-summary"
      data-testid="forecast-daily-summary"
    >
      <p>
        {dayOfTheWeek} <br />
        {monthAndDay}
      </p>
      <div className="forecast-daily-summary__icon" data-testid="forecast-icon">
        <WeatherIcon name="owm" iconId={icon} />
      </div>
      <p>
        {weatherDescription[0]} <br />
        {weatherDescription[1]}
      </p>
      <p>
        {temp}
        {tempUnits}
      </p>
      <button
        type="button"
        onClick={() => handleForecastSelect(dateTime)}
        className="forecast-daily-summary__more-details"
        disabled={selectedDate === dateTime}
      >
        More Details
      </button>
    </div>
  );
}

ForecastDailySummary.defaultProps = {
  selectedDate: 0,
};

ForecastDailySummary.propTypes = {
  dateTime: PropTypes.number.isRequired,
  weather: PropTypes.string.isRequired,
  icon: PropTypes.string.isRequired,
  temp: PropTypes.string.isRequired,
  handleForecastSelect: PropTypes.func.isRequired,
  selectedDate: PropTypes.number,
};

export default ForecastDailySummary;
