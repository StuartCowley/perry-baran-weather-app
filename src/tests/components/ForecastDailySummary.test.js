import React from "react";
import { render, fireEvent } from "@testing-library/react";
import ForecastDailySummary from "../../components/ForecastDailySummary";
import dateString from "../../helpers/dateString";

describe("ForecastDailySummary", () => {
  const validProps = {
    dateTime: 20,
    weather: "broken clouds",
    icon: "800",
    temp: "20",
    handleForecastSelect: jest.fn(),
  };
  const { dateTime, weather, icon, temp, handleForecastSelect } = validProps;
  let screen;

  beforeEach(() => {
    screen = render(
      <ForecastDailySummary
        dateTime={dateTime}
        weather={weather}
        icon={icon}
        temp={temp}
        handleForecastSelect={handleForecastSelect}
      />
    );
  });

  test("snapshot", () => {
    const { asFragment } = screen;

    expect(asFragment()).toMatchSnapshot();
  });

  test("correctly renders props", () => {
    const { getByText, getByTestId } = screen;
    const date = dateString(dateTime);
    const dayOfTheWeek = date.slice(0, 3);
    const monthAndDay = date.slice(3, 10);
    const description = weather.split(" ");

    expect(getByText(new RegExp(dayOfTheWeek))).toBeInstanceOf(
      HTMLParagraphElement
    );
    expect(getByText(new RegExp(monthAndDay))).toBeInstanceOf(
      HTMLParagraphElement
    );
    expect(getByTestId("forecast-icon")).toContainHTML("i");
    expect(getByText(new RegExp(description[0]))).toBeInstanceOf(
      HTMLParagraphElement
    );
    expect(getByText(new RegExp(description[1]))).toBeInstanceOf(
      HTMLParagraphElement
    );
    expect(getByText(new RegExp(temp))).toBeInstanceOf(HTMLParagraphElement);
  });

  test("button", () => {
    const { getByText } = screen;
    const button = getByText(/more details/i);

    expect(button).toBeInstanceOf(HTMLButtonElement);
    expect(handleForecastSelect).toBeCalledTimes(0);
    fireEvent.click(button);
    expect(handleForecastSelect).toBeCalledTimes(1);
    expect(handleForecastSelect).toBeCalledWith(dateTime);
  });
});
