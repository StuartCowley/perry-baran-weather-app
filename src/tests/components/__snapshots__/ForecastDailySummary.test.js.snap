// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastDailySummary snapshot 1`] = `
<DocumentFragment>
  <div
    class="forecast-daily-summary"
    data-testid="forecast-daily-summary"
  >
    <p>
      Thu 
      <br />
       Jan 01
    </p>
    <div
      class="forecast-daily-summary__icon"
      data-testid="forecast-icon"
    >
      <i
        class="wi wi-day-sunny"
        name="owm"
      />
    </div>
    <p>
      broken 
      <br />
      clouds
    </p>
    <p>
      20Â°C
    </p>
    <button
      class="forecast-daily-summary__more-details"
      type="button"
    >
      More Details
    </button>
  </div>
</DocumentFragment>
`;
